### ISSUE TITLE
SliceableIteratable in scikit-image?

### ISSUE THREAD

**Issue by danielballan** [soft-matter/pims#162]
@nkeim @caspervdw @tacaswell:

I'm at SciPy 2015 working next to @stefanv. He would like me to refactor scikit-image's `ImageCollection` to behave like pims objects, backed by `SliceableIterable`. After some discussion with @JDWarner, we concluded that the best way to start is to make a very light standalone package that both pims and scikit-image can depend upon, in the pattern of [cycler](https://github.com/matplotlib/cycler), [history](https://github.com/Nikea/history), and other simple utilities. This seems better than copying code and maintaining it in two places.

To be clear, I think pims has a good niche as a user-facing package, and this is not necessarily a step toward pushing pims wholly into scikit-image. It spreads the pims-like interface as a common pattern in the scipy stack.

While this would technically add a new dependency to pims, it wouldn't be a user-facing one: the user would not normally import it, and there will be never installation issues because `sliceable_iterable` depends only on core Python and `six`.

Questions:
1. Are you comfortable with this direction?
2. If so, would you be willing to be co-owners of a new `sliceable_iterable` repo? It is important to scikit-image that there be a couple people with permissions to the repo and PyPI.

**Comment by tacaswell** [comment]
And if we dropped 2 support it would have no dependencies!

I am :+1: and would willing to be a co-owner.

**Comment by nkeim** [comment]
Yes to both! Not sure what I bring to this except a shared enthusiasm for this pattern. Started wondering ~5 years ago why this wasn't in the standard library.

Nathan

On Jul 12, 2015, at 2:09 PM, Dan Allan <notifications@github.com<mailto:notifications@github.com>> wrote:


@nkeim<https://github.com/nkeim> @caspervdw<https://github.com/caspervdw> @tacaswell<https://github.com/tacaswell>:

I'm at SciPy 2015 working next to @stefanv<https://github.com/stefanv>. He would like me to refactor scikit-image's ImageCollection to behave like pims objects, backed by SliceableIterable. After some discussion with @JDWarner<https://github.com/JDWarner>, we concluded that the best way to start is to make a very light standalone package that both pims and scikit-image can depend upon, in the pattern of cycler<https://github.com/matplotlib/cycler>, history<https://github.com/Nikea/history>, and other simple utilities. This seems better than copying code and maintaining it in two places.

To be clear, I think pims has a good niche as a user-facing package, and this is not necessarily a step toward pushing pims wholly into scikit-image. It spreads the pims-like interface as a common pattern in the scipy stack.

While this would technically add a new dependency to pims, it wouldn't be a user-facing one: the user would not normally import it, and there will be never installation issues because sliceable_iterable depends only on core Python and six.

Questions:
1. Are you comfortable with this direction?
2. If so, would you be willing to be co-owners of a new sliceable_iterable repo? It is important to scikit-image that there be a couple people with permissions to the repo and PyPI.

**Comment by caspervdw** [comment]
Seems to me as a good direction! I am also willing to monitor the new repo.

What about the name, is `sliceable_iterable` the definite one? I always found it a bit long and difficult to read.

**Comment by nkeim** [comment]
LazyList

On Jul 13, 2015, at 12:55 AM, Casper van der Wel <notifications@github.com<mailto:notifications@github.com>> wrote:


Seems to me as a good direction! I am also willing to monitor the new repo.

What about the name, is sliceable_iterable the definite one? I always found it a bit long and difficult to read.

-

**Comment by caspervdw** [comment]
Reslicable

**Comment by ericdill** [comment]
Sliceable + Iterator --> Slicerator

**Comment by danielballan** [comment]
I like Slicerator. OK with everyone else?

P.S. LazyList is nice and simple, but in a way it's problematic because we don't support the list interface (e.g., `.extend`) like [this package of the same name](https://github.com/ryanhaining/lazylist) does.

**Comment by nkeim** [comment]
Ha! I was starting to like Slicerator better anyway.

On Jul 14, 2015, at 8:12 AM, Dan Allan <notifications@github.com<mailto:notifications@github.com>> wrote:


I like Slicerator. OK with everyone else?

P.S. LazyList is nice and simple, but in a way it's problematic because we don't support the list interface (e.g., .extend) like this package of the same name<https://github.com/ryanhaining/lazylist> does.

-

**Comment by caspervdw** [comment]
:+1: for Slicerator!

**Comment by JDWarner** [comment]
Slicerator gets my vote! As I was the one to ask for multiple owners, I'd
be happy to share that responsibility.

**Comment by stefanv** [comment]
Thanks for pushing this forward!

**Comment by danielballan** [comment]
Slicerator is live! See https://github.com/soft-matter/slicerator. It's on PyPI, and a conda package is on soft-matter's binstar channel.

There's a decent README but it could do with more examples. I tagged 0.9.x; I am to tag 1.0 right before we merge a refactored `ImageCollection` into scikit-image. This leaves us some room to rethink things until then.
